keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,1161 @@
+
+Additionally, this proposal is squarely focused on solving the needs of the
+Cluster Administrator. It is not intended to solve:
+1. Error reporting for Network Policy
+2. Kubernetes Node policies
+3. New policy selectors (services, service accounts, etc.)
+4. Cluster external traffic access control.
+
+## Proposal
+
+In order to achieve the two primary broad use cases for a cluster admin to
+secure K8s clusters, we propose to introduce the following resource
+under `netpol.networking.k8s.io` API group:
+- ClusterNetworkPolicy
+
+### ClusterNetworkPolicy resource
+
+A ClusterNetworkPolicy (CNP) resource will help the administrators set strict
+security rules for the cluster, i.e. a developer CANNOT override these rules
+by creating NetworkPolicies that applies to the same workloads as the
+ClusterNetworkPolicy does.
+
+#### Actions
+
+Unlike the NetworkPolicy resource in which each rule represents an allowed
+traffic, ClusterNetworkPolicy will enable administrators to set `Pass`,
+`Deny` or `Allow` as the action of each rule. ClusterNetworkPolicy rules should
+be read as-is, i.e. there will not be any implicit isolation effects for the Pods
+selected by the ClusterNetworkPolicy, as opposed to what NetworkPolicy rules imply.
+
+- Pass: Traffic that matches a `Pass` rule will skip all rules from all CNPs and instead
+  be enforced by the K8s NetworkPolicies. If there is no K8s NetworkPolicy rule match,
+  traffic will be allowed.
+- Deny: Traffic that matches a `Deny` rule will be dropped. This is a terminal action.
+- Allow: Traffic that matches an `Allow` rule will be allowed. This is a terminal action.
+
+ClusterNetworkPolicy `Deny` rules are useful for administrators to explicitly
+block traffic from malicious clients, or workloads that poses security risks.
+Those traffic restrictions can only be lifted once the `Deny` rules are deleted
+or modified by the admin.
+
+On the other hand, the `Allow` rules can be used to call out traffic in the cluster
+that needs to be allowed for certain components to work as expected (egress to
+CoreDNS for example). Those traffic should not be blocked when developers apply
+NetworkPolicy to their Namespaces which isolates the workloads.
+
+ClusterNetworkPolicy `Pass` rules allows an admin to delegate security posture for
+certain traffic to the Namespace owners. For example, allowing intra-tenant traffic
+can be delegated to tenant admins explicitly by the cluster admin with the use of `pass`
+rules.
+
+#### Priority
+
+The policy instances will be ordered based on the numeric priority assigned to the
+CNP. A lower number corresponds to a higher order. For example, CNP with `priority` set
+to "0" will be considered as the highest priority policy, while a large number, i.e.
+999, will be considered as low priority. Any numeric-valued priority, however, will have
+higher precedence over the namespaced NetworkPolicy instances in the cluster.
```

Comment:
I tend to lean towards keeping this simple and having 0 be the lowest 999 be the highest.... The main goal with flipping this was so that it would be easier to write  an "absolute" CNP... I.e I know if I have a priority 0 CNP to deny traffic to some pods that it will never be overridden

Reply:
For what it is worth, in Calico we call this field `order`, to indicate that policies are applied in order - i.e. 0 before 1, etc. I'm personally in favor of using that model here (independent of the name) - 0 is the first applied, and larger numbers have a lower priority.

What's the reason for choosing 999 as the highest? Seems arbitrary, and could be limiting for larger / more complex operators who might want to further build an ordering scheme within the allowed range. I'd suggest just allowing a uint32 here.

Reply:
I like the idea of using `order` as well! @caseydavenport The 999 limit was to ensure portability of CNP's between clusters, since some CNI's may only support a specific number of policies (for example OVN ACLs only support prioities between between 0 and 32767) however we assumed all CNIs should be able to support 1000 unique priorities.  Also we allow up to 100 rules within each policy so we figured that there should be plenty of space for admins to express their intent 

Lastly we believe there shouldn't be anywhere close to that many CNPs deployed in a cluster because typically there is only a single or small group of admins who would be writing/ managing those.

Reply:
+1 to Dan's suggestion on usage of words (in his edits) .. 

@caseydavenport @astoycos  I personally do not prefer 999(or any other arbitrary number) as highest. Would continue to suggest we keep 0 as the policy which takes highest precedence and so on..

Reply:
I lean towards 0 being the lowest importance, too.  I like "order" as a  clarifying name.  It also lets us use 0 as a  special value (baseline).